src/Nethermind/Nethermind.Core/Caching/NonBlockingLruCache.cs

```diff
@@ -0,0 +1,266 @@
+
+        private LinkedListNode<LruCacheItem>? _leastRecentlyUsed;
+        private CancellationTokenSource _cts;
+
+        private long _toProcess;
+        private int _processingLru;
+
+        public NonBlockingLruCache(int maxCapacity, int startCapacity, string name)
+        {
+            // Both types of concurrent dictionary heavily allocate if key is larger than pointer size.
+            ArgumentOutOfRangeException.ThrowIfGreaterThan(Unsafe.SizeOf<TKey>(), IntPtr.Size);
+
+            ArgumentOutOfRangeException.ThrowIfLessThan(maxCapacity, 1);
+
+            _maxCapacity = maxCapacity;
+            _accesses = new ConcurrentQueue<LinkedListNode<LruCacheItem>>();
+            _cacheMap = typeof(TKey) == typeof(byte[])
+                ? new ConcurrentDictionary<TKey, LinkedListNode<LruCacheItem>>((IEqualityComparer<TKey>)Bytes.EqualityComparer)
+                : new ConcurrentDictionary<TKey, LinkedListNode<LruCacheItem>>();
+            _cts = new CancellationTokenSource();
+        }
+
+        public NonBlockingLruCache(int maxCapacity, string name)
+            : this(maxCapacity, 0, name)
+        {
+        }
+
+        public void Clear()
+        {
+            // Signal background thread to stop
+            _cts.Cancel();
+            // Signal background thread to stop (two ways)
+            _leastRecentlyUsed = null;
+            _accesses.Clear();
+            _cacheMap.Clear();
+            _cts = new CancellationTokenSource();
+            // reclear incase background thread set it before accesses were cleared
+            _leastRecentlyUsed = null;
+        }
+
+        public TValue Get(TKey key) => TryGet(key, out TValue value) ? value : default!;
+
+        public bool TryGet(TKey key, out TValue value)
+        {
+            if (_cacheMap.TryGetValue(key, out LinkedListNode<LruCacheItem>? node))
+            {
+                value = node!.Value.Value;
+                ScheduleLruAccounting(node);
+                return true;
+            }
+
+            value = default!;
+            return false;
+        }
+
+        public void Set(TKey key, TValue val)
+        {
+            if (val is null)
+            {
+                Delete(key);
+            }
+
+            LinkedListNode<LruCacheItem> node = _cacheMap.AddOrUpdate(
+                key,
+                (k, v) => new(new(k, v)),
+                (k, node, v) =>
+                {
+                    node.Value.Value = val;
+                    return node;
+                },
+                val);
+
+            ScheduleLruAccounting(node, isGet: false);
+        }
+
+        bool ICache<TKey, TValue>.Set(TKey key, TValue val) => SetResult(key, val);
+
+        public bool SetResult(TKey key, TValue val)
+        {
+            if (val is null)
+            {
+                return Delete(key);
+            }
+
+            bool exists = false; // Captured by lambda
+            LinkedListNode<LruCacheItem> node = _cacheMap.AddOrUpdate(
```

Comment:
Do we always allocate an item on non-existence of the key here and not reuse the recent as thsi PR separates this two? Just asking, can be by design.